### PR TITLE
[SPARK-44524][BUILD] Balancing pyspark-pandas-connect and pyspark-pandas-slow-connect GA testing time

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to balancing `pyspark-pandas-connect` and `pyspark-pandas-slow-connect` GA testing time.

### Why are the changes needed?
After pr: https://github.com/apache/spark/pull/42146, the difference in testing time between `pyspark-pandas-connect` and `pyspark-pandas-slow-connect` is a bit significant, which affects the overall running time. In order to make GA operation more efficient and stable.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Manually monitor GA.

### CODE REVIEW

**Comment by LuciferYang** [comment]
image

It seems that splitting the test task is useful. Do we have to split it into 3?

**Comment by LuciferYang** [comment]
cc @HyukjinKwon @zhengruifeng FYI

**Review by HyukjinKwon — DISMISSED** [review]

**Review by zhengruifeng — DISMISSED** [review]
LGTM if CI pass

**Comment by panbingkun** [comment]
Maybe splitting into three is a stable choice. Let's try splitting into two first. Yesterday, splitting into three was successful.

**Comment by LuciferYang** [comment]
If dividing into two groups is still easy to fail, I do not oppose splitting into three groups :)

**Comment by zhengruifeng** [comment]
> If dividing into two groups is still easy to fail, I do not oppose splitting into three groups :)

If so, I suggest we'd better find better names for the groups (let's also take existing `pyspark_pandas_connect` into account), maybe `pyspark_pandas_connect_part_xxx`

**Comment by panbingkun** [comment]
> If so, I suggest we'd better find better names for the groups (let's also take existing `pyspark_pandas_connect` into account), maybe `pyspark_pandas_connect_part_xxx`

Ok, let me try again.

**Comment by zhengruifeng** [comment]
I think it is fine to split `pyspark-pandas-slow-connect` which is slow.

not sure whether they are due to the same cause, but I also see `pyspark-sql`, `pyspark-pandas-slow` failed for some reason (e.g. https://github.com/apache/spark/actions/runs/5651977344/job/15310946651)

It's kind of dangerous, since the GA recognized them as `skip` and made CI succeeded. 
image

**Comment by LuciferYang** [comment]
> It's kind of dangerous, since the GA recognized them as skip and made CI succeeded.

...

**Comment by panbingkun** [comment]
`pyspark-pandas-connect-part-3` failed to run in the final stage with the following error message:
image

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon does it mean the packaging test itself fails for some reason

**Comment by panbingkun** [comment]
Let's try to disable `run_python_packaging_tests`.

**Comment by panbingkun** [comment]
It succeeded.

**Comment by zhengruifeng** [comment]
@panbingkun yes, it seems the python packaging test is the blocker,  please hold on this PR, let's fix it in https://github.com/apache/spark/pull/42146 first.

**Comment by zhengruifeng** [comment]
@panbingkun after we disable packaging tests in `pyspark-pandas-connect` and `pyspark-pandas-slow-connect`, the `pyspark-pandas-connect` took 1h, and `pyspark-pandas-slow-connect` took 2h.
I think it is not worthwhile to split `pyspark-pandas-slow-connect` any more, instead I suggest rebalancing them by moving some tests from `pyspark-pandas-slow-connect` to `pyspark-pandas-connect`.

**Comment by panbingkun** [comment]
> @panbingkun after we disable packaging tests in `pyspark-pandas-connect` and `pyspark-pandas-slow-connect`, in https://github.com/zhengruifeng/spark/actions/runs/5664781463/job/15348941550 , the `pyspark-pandas-connect` took 1h, and `pyspark-pandas-slow-connect` took 2h. I think it is not worthwhile to split `pyspark-pandas-slow-connect` any more, instead I suggest rebalancing them by moving some tests from `pyspark-pandas-slow-connect` to `pyspark-pandas-connect`.

Ok,  let me turn this PR into the logic of rebalancing? 😄

**Comment by zhengruifeng** [comment]
> > @panbingkun after we disable packaging tests in `pyspark-pandas-connect` and `pyspark-pandas-slow-connect`, in https://github.com/zhengruifeng/spark/actions/runs/5664781463/job/15348941550 , the `pyspark-pandas-connect` took 1h, and `pyspark-pandas-slow-connect` took 2h. I think it is not worthwhile to split `pyspark-pandas-slow-connect` any more, instead I suggest rebalancing them by moving some tests from `pyspark-pandas-slow-connect` to `pyspark-pandas-connect`.
> 
> Ok, let me turn this PR into the logic of rebalancing? 😄

yeah, many thanks!

**Comment by panbingkun** [comment]
@HyukjinKwon @zhengruifeng @LuciferYang 
In addition, there may be a potential fix point needed in `run_python_packaging_tests ` testing, as it may affect the final release as `2023-09-26` time approaches.
Are we submitting a PR separately? Or did you just make it on this PR?
I prefer a separate PR because it may require backport to branch-3.3 & branch-3.4.
image

**Comment by zhengruifeng** [comment]
@panbingkun I see, what about fixing it after 3.5 is released?

**Comment by panbingkun** [comment]
> @panbingkun I see, what about fixing it after 3.5 is released?

Okay, fine to me.

**Comment by zhengruifeng** [comment]
please update this PR to latest master, since we just enabled > 100 test cases in https://github.com/apache/spark/commit/9b40c0cdc5cf8e24e66f7ee8e122702d3f157291 and https://github.com/apache/spark/commit/185a0a5a23958676e4236eaf9e4d78cdfd2dd2d7

**Comment by panbingkun** [comment]
> please update this PR to latest master, since we just enabled > 100 test cases in [9b40c0c](https://github.com/apache/spark/commit/9b40c0cdc5cf8e24e66f7ee8e122702d3f157291) and [185a0a5](https://github.com/apache/spark/commit/185a0a5a23958676e4236eaf9e4d78cdfd2dd2d7)

Okay, done.

**Comment by panbingkun** [comment]
@zhengruifeng 
After this modification, the testing time of `pyspark-pandas-connect` and `pyspark-pandas-slow-connect` has been basically balanced.

1.pyspark-pandas-connect
image

2.pyspark-pandas-slow-connect
image

File: dev/sparktestsupport/modules.py [hunk 1]
```diff
@@ -984,6 +967,23 @@ def __hash__(self):
...
         "pyspark.pandas.tests.connect.computation.test_parity_pivot",
+    ],
+    excluded_python_implementations=[
+        "PyPy"  # Skip these tests under PyPy since they require numpy, pandas, and pyarrow and
+        # they aren't available there
+    ],
+)
+
+
+# This module should contain the same test list with 'pyspark_pandas_slow' for maintenance.
```
Comment (zhengruifeng) [inline dev/sparktestsupport/modules.py:978]:
```suggestion
```

This comment is no longer the truth

File: dev/sparktestsupport/modules.py [hunk 2]
```diff
@@ -984,6 +967,23 @@ def __hash__(self):
...
+pyspark_pandas_slow_connect = Module(
+    name="pyspark-pandas-slow-connect",
+    dependencies=[pyspark_connect, pyspark_pandas_slow],
```
Comment (zhengruifeng) [inline dev/sparktestsupport/modules.py:981]:
```suggestion
    dependencies=[pyspark_connect, pyspark_pandas, pyspark_pandas_slow],
```

`pyspark_pandas_connect` should have the same dependencies now.

File: dev/sparktestsupport/modules.py [hunk 3]
```diff
@@ -984,6 +967,22 @@ def __hash__(self):
...
+    excluded_python_implementations=[
+        "PyPy"  # Skip these tests under PyPy since they require numpy, pandas, and pyarrow and
+        # they aren't available there
+    ],
+)
+
+
+pyspark_pandas_slow_connect = Module(
+    name="pyspark-pandas-slow-connect",
+    dependencies=[pyspark_connect, pyspark_pandas, pyspark_pandas_slow],
```
Comment (zhengruifeng) [inline dev/sparktestsupport/modules.py:980]: should also set `dependencies=[pyspark_connect, pyspark_pandas, pyspark_pandas_slow],` in the above `pyspark_pandas_connect` module

**Comment by panbingkun** [comment]
Let me double check the testing time this time, it looks quite balanced:
1.pyspark-pandas-connect
image

2.pyspark-pandas-slow-connect
image

File: dev/sparktestsupport/modules.py (hunk 3 above)
Comment (panbingkun) [reply dev/sparktestsupport/modules.py:980]: Done

**Comment by panbingkun** [comment]
Third run cost:
1.pyspark-pandas-connect
image

2.pyspark-pandas-slow-connect
image

**Review by HyukjinKwon — APPROVED** [review]
I am good with this. cc @itholic, @ueshin and @xinrong-meng

**Review by itholic — APPROVED** [review]

**Comment by zhengruifeng** [comment]
thanks, merged to master
